File: agora/plugin/server/skill_law.go
package main

import (
	"encoding/json"
	"fmt"
	"strings"
)

// skill_law, enforced SERVER-SIDE (authoritative). Mirrors the static checks of
// connector/laws/skill_law.py. The server NEVER runs a client-supplied selfcheck
// (no executing client code on the server) — the owner's connector runs selfcheck
// on its own host; the room enforces structure/contract/credentials/compat.
// See docs/laws/skill_law.md and docs/QUALITY-BAR.md §1 (never trust the client).

var skillInputTypes = map[string]bool{
	"string": true, "int": true, "float": true, "bool": true, "list": true, "object": true,
}
var skillSecretKeys = map[string]bool{
	"value": true, "secret": true, "password": true, "pass": true,
	"token": true, "key": true, "apikey": true, "api_key": true,
}
var skillOSes = []string{"windows", "macos", "linux"}

type skillReason struct {
	Clause string `json:"clause"`
	OS     string `json:"os,omitempty"`
	Detail string `json:"detail"`
}

type skillCompat struct {
	Status string `json:"status"`
	Detail string `json:"detail"`
}

type skillVerdict struct {
	Skill   string                 `json:"skill"`
	Verdict string                 `json:"verdict"`
	Reasons []skillReason          `json:"reasons"`
	Compat  map[string]skillCompat `json:"compat"`
}

func (v *skillVerdict) fail(clause, onOS, detail string) {
	v.Reasons = append(v.Reasons, skillReason{Clause: clause, OS: onOS, Detail: detail})
}

func nonEmptyStr(v interface{}) bool {
	s, ok := v.(string)
	return ok && strings.TrimSpace(s) != ""
}

func strOf(v interface{}) string { s, _ := v.(string); return s }

// gateSkillManifest validates one raw manifest and returns the authoritative verdict.
func gateSkillManifest(raw json.RawMessage) skillVerdict {
	v := skillVerdict{Skill: "<unnamed>", Verdict: "REJECT", Reasons: []skillReason{}, Compat: map[string]skillCompat{}}

	var any interface{}
	if err := json.Unmarshal(raw, &any); err != nil {
		v.Skill = "<unparseable>"
		v.fail("PARSE_ERROR", "", "invalid JSON: "+err.Error())
		return v
	}
	m, ok := any.(map[string]interface{})
	if !ok {
		v.fail("SCHEMA", "", "manifest is not an object")
		return v
	}

	if nonEmptyStr(m["name"]) {
		v.Skill = m["name"].(string)
	} else {
		v.fail("NAME", "", "missing or empty 'name'")
	}
	if !nonEmptyStr(m["version"]) {
		v.fail("VERSION", "", "missing or empty 'version'")
	}

	// description
	if desc, ok := m["description"].(map[string]interface{}); !ok {
		v.fail("DESCRIPTION", "", "missing 'description' object")
	} else {
		for _, f := range []string{"what", "when_to_use", "not_for"} {
			if !nonEmptyStr(desc[f]) {
				v.fail("DESCRIPTION_INCOMPLETE", "", "description."+f+" missing/empty")
			}
		}
	}

	// inputs (typed)
	if inputs, ok := m["inputs"].([]interface{}); !ok {
		v.fail("INPUTS", "", "missing 'inputs' list")
	} else {
		for i, ri := range inputs {
			inp, ok := ri.(map[string]interface{})
			if !ok {
				v.fail("INPUT_SHAPE", "", fmt.Sprintf("inputs[%d] not an object", i))
				continue
			}
			if !nonEmptyStr(inp["name"]) {
				v.fail("INPUT_SHAPE", "", fmt.Sprintf("inputs[%d] missing name", i))
			}
			if !skillInputTypes[strOf(inp["type"])] {
				v.fail("INPUT_TYPE", "", fmt.Sprintf("inputs[%d] type %q invalid", i, strOf(inp["type"])))
			}
			if r, present := inp["required"]; present {
				if _, isBool := r.(bool); !isBool {
					v.fail("INPUT_SHAPE", "", fmt.Sprintf("inputs[%d].required must be bool", i))
				}
			}
		}
	}

	// outputs (typed)
	if outputs, ok := m["outputs"].([]interface{}); !ok {
		v.fail("OUTPUTS", "", "missing 'outputs' list")
	} else {
		for i, ro := range outputs {
			out, ok := ro.(map[string]interface{})
			if !ok || !nonEmptyStr(out["name"]) || !skillInputTypes[strOf(out["type"])] {
				v.fail("OUTPUT_SHAPE", "", fmt.Sprintf("outputs[%d] needs name + valid type", i))
			}
		}
	}

	// errors (enumerated; no silent failure)
	if errs, ok := m["errors"].([]interface{}); !ok || len(errs) == 0 {
		v.fail("NO_ERROR_CONTRACT", "", "must enumerate >=1 typed error (no silent failure)")
	} else {
		for i, re := range errs {
			e, ok := re.(map[string]interface{})
			if !ok || !nonEmptyStr(e["code"]) || !nonEmptyStr(e["when"]) {
				v.fail("ERROR_SHAPE", "", fmt.Sprintf("errors[%d] needs code + when", i))
			}
		}
	}

	// workflow (end-to-end)
	if wf, ok := m["workflow"].(map[string]interface{}); !ok {
		v.fail("WORKFLOW", "", "missing 'workflow' object")
	} else {
		if steps, ok := wf["steps"].([]interface{}); !ok || len(steps) == 0 {
			v.fail("WORKFLOW", "", "workflow.steps must be a non-empty list")
		}
		for _, f := range []string{"preconditions", "postconditions"} {
			if _, ok := wf[f].([]interface{}); !ok {
				v.fail("WORKFLOW", "", "workflow."+f+" must be a list")
			}
		}
	}

	// failure semantics
	if fl, ok := m["failure"].(map[string]interface{}); !ok {
		v.fail("FAILURE", "", "missing 'failure' object")
	} else {
		if fl["blast_radius"] == nil || strings.TrimSpace(fmt.Sprintf("%v", fl["blast_radius"])) == "" {
			v.fail("FAILURE", "", "failure.blast_radius required (Art. IV)")
		}
		if idem, present := fl["idempotent"]; present {
			if _, isBool := idem.(bool); !isBool {
				v.fail("FAILURE", "", "failure.idempotent must be bool")
			}
		}
	}

	// credentials: ids only, never values (Art. II)
	if credsRaw, present := m["credentials"]; present {
		creds, ok := credsRaw.([]interface{})
		if !ok {
			v.fail("CREDENTIALS", "", "'credentials' must be a list")
		} else {
			for i, rc := range creds {
				c, ok := rc.(map[string]interface{})
				if !ok {
					continue
				}
				for k := range c {
					if skillSecretKeys[strings.ToLower(k)] {
						v.fail("CREDS_IN_MANIFEST", "", fmt.Sprintf("credentials[%d] embeds %q; use a ref", i, k))
					}
				}
				if !nonEmptyStr(c["ref"]) {
					v.fail("CREDENTIALS", "", fmt.Sprintf("credentials[%d] must carry a 'ref'", i))
				}
			}
		}
	}

	// compatibility: per-OS status (never rejects; missing OS = graceful incompat)
	osSupport, _ := m["os_support"].(map[string]interface{})
	for _, o := range skillOSes {
		spec, _ := osSupport[o].(map[string]interface{})
		switch {
		case spec != nil && nonEmptyStr(spec["resolve"]):
			v.Compat[o] = skillCompat{Status: "ok", Detail: strOf(spec["resolve"])}
		case spec != nil && nonEmptyStr(spec["graceful"]):
			v.Compat[o] = skillCompat{Status: "graceful", Detail: strOf(spec["graceful"])}
		default:
			v.Compat[o] = skillCompat{Status: "incompatible",
				Detail: fmt.Sprintf("incompatible on %s - install support or declare a 'resolve' to enable", o)}
		}
	}

	if len(v.Reasons) == 0 {
		v.Verdict = "ADMIT"
	}
	return v
}

File: agora/connector/laws/skill_law.py
"""skill_law — the gate that ADMITS or hard-REJECTS a skill.

Agora Law (see docs/laws/skill_law.md). Pure stdlib, cross-OS.
A skill is ADMITTED only if it has ZERO violations; otherwise REJECTED with a
structured list of reasons (which clause failed, on which OS). No partial admission.

Register-time check (what runs here):
  - static: the manifest conforms to the contract (typed inputs/outputs/errors,
    complete description, declared workflow, no creds in the manifest).
  - compatibility: every OS (windows/macos/linux) declares a resolution OR a
    graceful refusal -> nothing can hard-crash (Constitution Art. I).
  - dynamic dry-run: if the skill ships a selfcheck, run it on the HOST OS;
    the other two OSes are dry-resolved (declared-resolution present).
Graduate-time (real pass on all 3 OS) is a separate step; see check_graduation().
"""
import json
import os
import platform
import subprocess

OSES = ("windows", "macos", "linux")
INPUT_TYPES = {"string", "int", "float", "bool", "list", "object"}
# keys in a credential entry that would mean a secret VALUE is embedded (Art. II)
SECRET_KEYS = {"value", "secret", "password", "pass", "token", "key", "apikey", "api_key"}


def host_os():
    s = platform.system().lower()
    return {"darwin": "macos", "windows": "windows", "linux": "linux"}.get(s, s)


class Verdict:
    def __init__(self, name):
        self.name = name
        self.reasons = []  # [{clause, os, detail}]  -> hard-reject if any
        self.compat = {}   # {os: {status: ok|graceful|incompatible, detail}}  -> never rejects

    def fail(self, clause, detail, on_os=None):
        self.reasons.append({"clause": clause, "os": on_os, "detail": detail})

    @property
    def admit(self):
        return not self.reasons

    def to_dict(self):
        return {
            "skill": self.name,
            "verdict": "ADMIT" if self.admit else "REJECT",
            "reasons": self.reasons,
            "compat": self.compat,
        }


def _is_nonempty_str(x):
    return isinstance(x, str) and x.strip() != ""


def check_manifest(manifest, run_selfcheck=True, host=None):
    """Validate one parsed manifest dict. Returns a Verdict."""
    host = host or host_os()
    name = manifest.get("name") if isinstance(manifest, dict) else None
    v = Verdict(name if _is_nonempty_str(name) else "<unnamed>")

    if not isinstance(manifest, dict):
        v.fail("SCHEMA", "manifest is not an object")
        return v

    # --- identity ---
    if not _is_nonempty_str(manifest.get("name")):
        v.fail("NAME", "missing or empty 'name'")
    if not _is_nonempty_str(manifest.get("version")):
        v.fail("VERSION", "missing or empty 'version'")

    # --- description (must let an agent decide WHEN to use it) ---
    desc = manifest.get("description")
    if not isinstance(desc, dict):
        v.fail("DESCRIPTION", "missing 'description' object")
    else:
        for field in ("what", "when_to_use", "not_for"):
            if not _is_nonempty_str(desc.get(field)):
                v.fail("DESCRIPTION_INCOMPLETE", f"description.{field} missing/empty")

    # --- inputs: typed ---
    inputs = manifest.get("inputs")
    if not isinstance(inputs, list):
        v.fail("INPUTS", "missing 'inputs' list")
    else:
        for i, inp in enumerate(inputs):
            if not isinstance(inp, dict):
                v.fail("INPUT_SHAPE", f"inputs[{i}] not an object"); continue
            if not _is_nonempty_str(inp.get("name")):
                v.fail("INPUT_SHAPE", f"inputs[{i}] missing name")
            t = inp.get("type")
            if t not in INPUT_TYPES:
                v.fail("INPUT_TYPE", f"inputs[{i}] type {t!r} not in {sorted(INPUT_TYPES)}")
            if not isinstance(inp.get("required", False), bool):
                v.fail("INPUT_SHAPE", f"inputs[{i}].required must be bool")

    # --- outputs: typed ---
    outputs = manifest.get("outputs")
    if not isinstance(outputs, list):
        v.fail("OUTPUTS", "missing 'outputs' list")
    else:
        for i, out in enumerate(outputs):
            if not (isinstance(out, dict) and _is_nonempty_str(out.get("name")) and out.get("type") in INPUT_TYPES):
                v.fail("OUTPUT_SHAPE", f"outputs[{i}] needs name + valid type")

    # --- errors: enumerated, no silent failure ---
    errors = manifest.get("errors")
    if not isinstance(errors, list) or len(errors) == 0:
        v.fail("NO_ERROR_CONTRACT", "must enumerate >=1 typed error (no silent failure)")
    else:
        for i, e in enumerate(errors):
            if not (isinstance(e, dict) and _is_nonempty_str(e.get("code")) and _is_nonempty_str(e.get("when"))):
                v.fail("ERROR_SHAPE", f"errors[{i}] needs code + when")

    # --- workflow: end-to-end ---
    wf = manifest.get("workflow")
    if not isinstance(wf, dict):
        v.fail("WORKFLOW", "missing 'workflow' object")
    else:
        if not isinstance(wf.get("steps"), list) or len(wf.get("steps") or []) == 0:
            v.fail("WORKFLOW", "workflow.steps must be a non-empty list")
        for field in ("preconditions", "postconditions"):
            if not isinstance(wf.get(field), list):
                v.fail("WORKFLOW", f"workflow.{field} must be a list")

    # --- compatibility: per-OS status (NEVER rejects; missing OS = graceful incompat, Art. I) ---
    oss = manifest.get("os_support") if isinstance(manifest.get("os_support"), dict) else {}
    for o in OSES:
        spec = oss.get(o)
        if isinstance(spec, dict) and _is_nonempty_str(spec.get("resolve")):
            v.compat[o] = {"status": "ok", "detail": spec["resolve"]}
        elif isinstance(spec, dict) and _is_nonempty_str(spec.get("graceful")):
            v.compat[o] = {"status": "graceful", "detail": spec["graceful"]}
        else:
            v.compat[o] = {"status": "incompatible",
                           "detail": f"incompatible on {o} - install support or declare a 'resolve' to enable"}

    # --- failure semantics ---
    fail_spec = manifest.get("failure")
    if not isinstance(fail_spec, dict):
        v.fail("FAILURE", "missing 'failure' object")
    else:
        if not _is_nonempty_str(str(fail_spec.get("blast_radius", ""))):
            v.fail("FAILURE", "failure.blast_radius required (Art. IV)")
        if not isinstance(fail_spec.get("idempotent", False), bool):
            v.fail("FAILURE", "failure.idempotent must be bool")

    # --- credentials: ids only, never values (Art. II) ---
    creds = manifest.get("credentials", [])
    if creds and not isinstance(creds, list):
        v.fail("CREDENTIALS", "'credentials' must be a list")
    else:
        for i, c in enumerate(creds or []):
            if isinstance(c, dict):
                leaked = SECRET_KEYS.intersection(k.lower() for k in c.keys())
                if leaked:
                    v.fail("CREDS_IN_MANIFEST", f"credentials[{i}] embeds {sorted(leaked)}; use a ref", )
                if not _is_nonempty_str(c.get("ref")):
                    v.fail("CREDENTIALS", f"credentials[{i}] must carry a 'ref'")

    # --- dynamic dry-run: run selfcheck on the host OS only ---
    sc = manifest.get("selfcheck")
    if run_selfcheck and isinstance(sc, dict) and isinstance(sc.get("cmd"), list) and sc["cmd"]:
        try:
            r = subprocess.run(sc["cmd"], capture_output=True, text=True,
                               timeout=int(sc.get("timeout", 30)))
            if r.returncode != 0:
                v.fail("SELFCHECK_FAILED", f"exit {r.returncode}: {(r.stderr or r.stdout)[:200]}", on_os=host)
        except subprocess.TimeoutExpired:
            v.fail("SELFCHECK_FAILED", "selfcheck timed out", on_os=host)
        except Exception as ex:  # noqa: BLE001
            v.fail("SELFCHECK_FAILED", f"could not run selfcheck: {ex}", on_os=host)

    return v


def check_file(path, run_selfcheck=True):
    """Validate a manifest file. JSON parse errors are themselves a REJECT."""
    name = os.path.splitext(os.path.basename(path))[0]
    try:
        with open(path, encoding="utf-8") as f:
            manifest = json.load(f)
    except json.JSONDecodeError as e:
        v = Verdict(name)
        v.fail("PARSE_ERROR", f"invalid JSON: {e}")
        return v
    return check_manifest(manifest, run_selfcheck=run_selfcheck)


def check_graduation():
    """Graduate = real pass on windows+macos+linux. Stub: wired to the CI matrix /
    the Test Ledger when we build skill graduation. Register-time uses check_manifest."""
    raise NotImplementedError("graduation runs the real 3-OS matrix; not part of register")
